### PR TITLE
Support for constructing `Boxes2D` from arrays of arbitrary formats

### LINKED ISSUE #3434: Make `Box2D` accept all of the old `RectFormat` alternatives
The construction of Boxes2D requires more conversion in user code than the old `log_rect` and `log_rects` due to dropping `RectFormat` alternatives. To start with it is much more common for neural network libraries to output detections in 4-value array form instead of the new 2x2 format. 

This issue is made even worse due to
- https://github.com/rerun-io/rerun/issues/3433

Current docs:
Screenshot 2023-09-25 at 09 06 57

At the very least we should update the docstrings to explain very clearly how to log rectangles in all the different format. Preferably we would also support letting the user log rects in 4xN formats as well, perhaps using some class method convenience constructors.

### PR DESCRIPTION
**Pull request by teh-cmc**
- Fixes #3434
- Required for #3382

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [ ] I have tested [demo.rerun.io](https://demo.rerun.io/pr/3533) (if applicable)

- [PR Build Summary](https://build.rerun.io/pr/3533)
- [Docs preview](https://rerun.io/preview/aa2e9160727ef85adfe47443c6de96bca9eb8144/docs) 
- [Examples preview](https://rerun.io/preview/aa2e9160727ef85adfe47443c6de96bca9eb8144/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

File: rerun_py/rerun_sdk/rerun/log_deprecated/rects.py [hunk 1]
```diff
@@ -180,16 +180,16 @@ def log_rects(
         half_sizes = rects[:, 2:4] / 2
         centers = rects[:, 0:2] + half_sizes
     elif rect_format == RectFormat.YXHW:
-        half_sizes = rects[:, 4:2] / 2
-        centers = rects[:, 2:0] + half_sizes
```
Comment (teh-cmc) [inline rerun_py/rerun_sdk/rerun/log_deprecated/rects.py:184]: `i0 >= i1` will always return an empty array no matter what... we just never tested this path.

File: rerun_py/rerun_sdk/rerun/archetypes/boxes2d_ext.py [hunk 2]
```diff
@@ -57,26 +87,70 @@ def __init__(
...
-            if half_sizes is None:
-                _send_warning("Cannot specify `mins` without `sizes` or `half_sizes`.", 1)
-                half_sizes = np.asarray([1, 1], dtype=np.float32)
-
-            mins = np.asarray(mins, dtype=np.float32)
-            half_sizes = np.asarray(half_sizes, dtype=np.float32)
-            centers = mins + half_sizes
+                _send_warning("Cannot specify both `array` and `centers` at the same time.", 1)
+
+            if np.any(array):
```
Comment (Wumpf) [inline rerun_py/rerun_sdk/rerun/archetypes/boxes2d_ext.py:100]: np.any is documented as 

> Test whether any array element along a given axis evaluates to True.

Don't get it, what does this do here then? Why don't you just always `np.asarray` the incoming array, I think that's what we do everywhere else?

File: rerun_py/rerun_sdk/rerun/log_deprecated/rects.py [hunk 3]
```diff
@@ -125,7 +125,7 @@ def log_rects(
 
     Colors should either be in 0-255 gamma space or in 0-1 linear space.
     Colors can be RGB or RGBA. You can supply no colors, one color,
-    or one color per point in a Nx3 or Nx4 numpy array.
+    or one color per point in a Nx3 or Nx4 numpy rects.
```
Comment (Wumpf) [inline rerun_py/rerun_sdk/rerun/log_deprecated/rects.py:128]: the colors isn't an "numpy rect" (nothing says "that many arrays per rect", it's just a single array with the given dimension)

File: rerun_py/rerun_sdk/rerun/log_deprecated/rects.py [hunk 4]
```diff
@@ -138,7 +138,7 @@ def log_rects(
     entity_path:
         Path to the rectangles in the space hierarchy.
     rects:
-        Nx4 numpy array, where each row is [x, y, w, h], or some format you pick with the `rect_format` argument.
+        Nx4 numpy rects, where each row is [x, y, w, h], or some format you pick with the `rect_format` argument.
```
Comment (Wumpf) [inline rerun_py/rerun_sdk/rerun/log_deprecated/rects.py:141]: similar, it's a single Nx4 numpy _array_. It's just that `N` is the number of rects. So maybe
```suggestion
        Nx4 numpy array for N rects, where each row is [x, y, w, h], or some format you pick with the `rect_format` argument.
```

**Review by Wumpf — APPROVED** [review]
some comments on comments otherwise lgtm
Thought we wanted this as a separate utility method now instead of adding more arguments? but I suppose this is actually more natural after all. Wondering if we can come up with a better name than 'array' - looks a bit weird to do`rr.Boxes2D(array=`, but I have no better idea either

**Comment by teh-cmc** [comment]
> Thought we wanted this as a separate utility method now instead of adding more arguments? 

I prefer the former, but right now we already do the latter in a bunch of places... while I despise combinatorial constructor, I just hate inconsistency even more :zany_face:

File: rerun_py/rerun_sdk/rerun/log_deprecated/rects.py (hunk 3 above)
Comment (teh-cmc) [reply rerun_py/rerun_sdk/rerun/log_deprecated/rects.py:128]: sed gone wrong

**Pull request merged**